code/controllers/autofire.dm

```diff
@@ -0,0 +1,264 @@
+	var/bucket_pos = BUCKET_POS(next_fire)
+
+	// Get local references to the subsystem's vars, faster than accessing on the datum
+	var/list/bucket_list = SSautomatedfire.bucket_list
+
+	// Attempt to get the head of the bucket
+	var/datum/component/automatedfire/bucket_head
+	if (bucket_pos > 0)
+		bucket_head = bucket_list[bucket_pos]
+
+	//Replace the bucket head if needed
+	if(bucket_head == src)
+		bucket_list[bucket_pos] = next
+	SSautomatedfire.bucket_count--
+	// Remove the shooter from the bucket, ensuring to maintain
+	// the integrity of the bucket's list if relevant
+	if(prev != next)
+		prev.next = next
+		next.prev = prev
+	else
+		prev?.next = null
+		next?.prev = null
+	prev = next = null
+
+///Handle the firing of the autofire component
+/datum/component/automatedfire/proc/process_shot()
+	return
+
+#undef BUCKET_LEN
+#undef BUCKET_POS
+
+/obj/structure/turret_debug
+	name = "debug turret"
+	///What kind of ammo it uses
+	var/datum/ammo/ammo
+	///Its target
+	var/atom/target 
+	///At wich rate it fires in ticks
+	var/firerate = 5
+
+/obj/structure/turret_debug/fast
+	name = "debug turret fast"
+	firerate = 1
+
+/obj/structure/turret_debug/slow
+	name = "debug turret slow"
+	firerate = 25
+
+/obj/structure/turret_debug/Initialize()
+	. = ..()
+	ammo = GLOB.ammo_list[/datum/ammo/xeno/acid]
+	target = locate(x+5, y, z)
+	AddComponent(/datum/component/automatedfire/automatic_shoot_at, firerate, ammo)
+	SEND_SIGNAL(src, COMSIG_AUTOMATIC_SHOOTER_START_SHOOTING_AT, target)
+
+/datum/component/automatedfire/automatic_shoot_at
+	///The target we are shooting at
+	var/atom/target
+	///The ammo we are shooting
+	var/datum/ammo/ammo
+	///The delay between each shot in ticks
+	var/shot_delay = 5
+	///If we are shooting
+	var/shooting = FALSE
+
+/datum/component/automatedfire/automatic_shoot_at/Initialize(_shot_delay, _ammo)
+	. = ..()
+	if(!isatom(parent))
+		return COMPONENT_INCOMPATIBLE
+	shooter = parent
+	shot_delay = _shot_delay
+	ammo = _ammo
+	RegisterSignal(parent, COMSIG_AUTOMATIC_SHOOTER_START_SHOOTING_AT, .proc/start_shooting)
+	RegisterSignal(parent, COMSIG_AUTOMATIC_SHOOTER_STOP_SHOOTING_AT, .proc/stop_shooting)
+
+///Signal handler for starting the autoshooting at something
+/datum/component/automatedfire/automatic_shoot_at/proc/start_shooting(datum/source, _target)
+	SIGNAL_HANDLER
+	target = _target
+	next_fire = world.time
+	if(!shooting)
+		shooting = TRUE
+		schedule_shot()
+
+
+///Signal handler for stoping the shooting
+/datum/component/automatedfire/automatic_shoot_at/proc/stop_shooting(datum/source)
+	SIGNAL_HANDLER
+	target = null
+	shooting = FALSE
+
+/datum/component/automatedfire/automatic_shoot_at/process_shot()
+	if(!shooting)
+		return AUTOFIRE_STOPPED_SHOOTING
+	var/obj/projectile/newshot = new(shooter.loc)
+	newshot.generate_bullet(ammo)
+	newshot.permutated += shooter
+	newshot.fire_at(target, shooter, null, ammo.max_range, ammo.shell_speed)
+	SEND_SIGNAL(shooter, COMSIG_AUTOMATIC_SHOOTER_SHOT_FIRED)
```

Comment:
This signal is not used anywhere?

Reply:
Yeah, i use it in the xenos turret and the gun refactoring. That's just for example purpose, cause that's how communication between the comp and the gun/turret is done